onnxruntime/python/onnxruntime_pybind_state.cc

```diff
@@ -279,19 +287,38 @@ void GetPyObjFromTensor(const Tensor& rtensor, py::object& obj, const DataTransf
 
   if (numpy_type != NPY_OBJECT) {
     //if it is not cpu tensor, need to copy to host
-    if (rtensor.Location().device.Type() != OrtDevice::CPU) {
-      if (!data_transfer_manager)
-        throw std::runtime_error("GetPyObjFromTensor: data transfer manager is needed to convert non-CPU tensor to numpy array");
+    auto device_type = rtensor.Location().device.Type();
+    if (device_type != OrtDevice::CPU) {
+      if (!data_transfer_manager && !mem_cpy_to_host_functions)
+        throw std::runtime_error(
+            "GetPyObjFromTensor: Either data transfer manager or a "
+            "function to copy data to the host is needed to convert non-CPU tensor to numpy array");
       static const OrtMemoryInfo cpu_alloc_info{onnxruntime::CPU, OrtDeviceAllocator};
-      std::vector<char> tensor_data_buffer{};
-      tensor_data_buffer.resize(rtensor.SizeInBytes());
-      ORT_THROW_IF_ERROR(CopyTensorDataToByteSpan(
-          *data_transfer_manager, rtensor, cpu_alloc_info, gsl::make_span(tensor_data_buffer)));
-      memcpy(outPtr, tensor_data_buffer.data(), dtype->Size() * shape.Size());
+
+      // Prefer DataTransferManager if available
+      if (data_transfer_manager) {
+        auto span = gsl::make_span<char>(reinterpret_cast<char*>(outPtr), dtype->Size() * shape.Size());
+        ORT_THROW_IF_ERROR(CopyTensorDataToByteSpan(
+            *data_transfer_manager, rtensor, cpu_alloc_info, span));
+      } else {
+        auto mem_cpy_to_host = mem_cpy_to_host_functions->find(device_type);
+
+        ORT_ENFORCE(mem_cpy_to_host != mem_cpy_to_host_functions->end(),
+                    "Unable to locate a function that can copy data to the host from the device");
+
+        ORT_ENFORCE(mem_cpy_to_host->second != 0,
+                    "No function that can copy data to the host from the device provided");
+
+        mem_cpy_to_host->second(outPtr, rtensor.DataRaw(), dtype->Size() * shape.Size());
+      }
+
     } else
       memcpy(outPtr, rtensor.DataRaw(dtype), dtype->Size() * shape.Size());
   } else {
     // Handle string type.
+    //if it is not cpu tensor, need to copy to host
+    ORT_ENFORCE(rtensor.Location().device.Type() == OrtDevice::CPU,
```

Comment:
Added this enforce so as to make sure this block is not entered for strings - Afaik - strings tensors are not supported in CUDA anyway - so we shouldn't hit this

Reply:
Comment on previous line is confusing as it says you need to copy to host, and this says we don't support copying to host. 

---
In reply to: [492465144](https://github.com/microsoft/onnxruntime/pull/5248#discussion_r492465144) [](ancestors = 492465144)

Reply:
You're right - fixed - thanks